### DISCUSSION TITLE
Padding around a widget

### DISCUSSION THREAD

**Discussion by John-Nagle** [emilk/egui#2588] [Q&A]
![bottombar](https://user-images.githubusercontent.com/942608/212520297-fdbac4ed-284b-4bf2-a1ed-8cb3a001b6da.png)

So I want to put padding space on all 4 sides of a set of widgets, then draw a round-cornered box around the set to indicate grouping. Like the one above.

I can get egui to do this, but it's really clunky. I'm doing this in a **TopBottomPanel::bottom().** So everything is aligned to the bottom of the window, and the pane has the height of the tallest item.

So the bottom pane code looks like this:
```
 egui::TopBottomPanel::bottom("bottom_panel")
                .frame(Frame::none().fill(TRANSLUCENT_GREY_COLOR32))
                .show(&ctx, |ui| {
                    ui.visuals_mut().widgets.inactive.bg_fill = egui::Color32::TRANSPARENT; // transparent button background
                    ui.with_layout(egui::Layout::bottom_up(egui::Align::LEFT), |ui| {
                        ui.horizontal(|ui| {
                            //  Basic camera controls -- orbit, zoom, move
                            ui.add_space(OUTLINE_BOX_ROUNDING); // allow extra space before next item ***TEMP***
                            self.panel_camera.draw(ui, state);
                            //  Movement controls - walk, run, fly, stand.
                            ui.add_space(OUTLINE_BOX_ROUNDING); // allow extra space before next item ***TEMP***
                            self.panel_movement.draw(ui, state);
                            //  Chat controls - input text box
                            ui.add_space(OUTLINE_BOX_ROUNDING); // allow extra space before next item ***TEMP***
                            self.panel_chat.draw(ui, state);
                        })                
                    });
                });
```
I have to ask for bottom-up alignment, or I get a huge space below the widgets.

Then each widget needs lots of vertical-horizontal switching to add padding on both axes on both sides. Here's the code for those two arrow widgets.

```
        ui.horizontal(|ui| {
            ui.add_space(OUTLINE_BOX_PADDING); // to allow for outline box.
            let cam_orbit_action = { 
                ui.with_layout(egui::Layout::bottom_up(egui::Align::LEFT), |ui| {
                    ui.add_space(OUTLINE_BOX_PADDING); // vertical space to allow for outline box.
                    let response = ui.add(&mut self.orbit_arrows);
                    let cam_orbit_action = self.orbit_arrows.decode_response(&response);
                    ui.add_space(OUTLINE_BOX_PADDING); // vertical space to allow for outline box.
                    cam_orbit_action
                }).inner
            };
```
Outside this, there's bottom-up layout. Then there's horizontal layout, to allow padding horizontally, Then a switch to vertical bottom up layout to allow adding bottom vertical padding. Then the widget. Then more padding above the widget.

Does it have to be this complicated? I'd really like to just be able to add a widget with padding. Is there an easier way?

The arrow widget code is [here](https://github.com/John-Nagle/ui-mock/blob/main/src/navarrows.rs) if anyone wants to see that.

**Comment by bglw** [comment]
Not entirely sure on the layout fronts, but this is how I currently achieve grouping a set of widgets in a layout, adding padding around them and adding a stroke around that:

```rust
ui.allocate_ui_with_layout(
    egui::vec2(ui.available_width(), 100.0),
    egui::Layout::top_down(egui::Align::LEFT),
    |ui| {
        Frame::none()
            .inner_margin(Margin::symmetric(12.0, 12.0))
            .rounding(8.0)
            .stroke(Stroke::new(1.0, Color32::GOLD))
            .show(ui, |ui| { /* widgets */ });
    },
);
```

**Reply by bglw** [reply]
For a lot of widgets, the internal padding can also be done a bit quicker with this above the `Frame`
```rust
ui.style_mut().spacing.item_spacing = egui::vec2(10.0, 10.0);
```

**Comment by John-Nagle** [comment]
I found that using "Frame" did what I needed. Still some problems with frame top and bottom padding, but I'm on an old version of egui ad will retest after Rend3 releases a new version that is compatible with egui 0.20. What I have now is below.

![bottombartest](https://user-images.githubusercontent.com/942608/213979834-cd490f99-e9bd-465d-b795-666762d54c22.png)
